File: leetcode/HouseRobber2.py
class Solution {
public:
    int rob(vector<int>& nums) {
        if(nums.empty()){
            return 0;
        }
        if(nums.size() <= 3){
            return *max_element(nums.begin(), nums.end());
        }
        vector<int> nums2 = nums;
        for(int i=nums.size()-4; i >= 0; i--){
            if(i+3 < nums.size()-1){
                nums[i] += max(nums[i+2], nums[i+3]);
            }
            else{
                nums[i] += nums[i+2];
            }
        }
        
        for(int i=nums2.size()-3; i>=1; i--){
            if(i+3 < nums2.size()){
                nums2[i] += max(nums2[i+2], nums2[i+3]);
            } 
            else{
                nums2[i] += nums2[i+2];
            }
        }
        return max({nums[0], nums[1], nums2[1], nums2[2]});
    }
};

File: leetcode/HouseRobber2.cpp
class Solution {
public:
    int rob(vector<int>& nums) {
        int n = nums.size();
        vector<int> dp1, dp2;
        if(n==1)
            return nums[0];
        if(n==2)
            return max(nums[0], nums[1]);
        dp1=nums;
        dp2=nums;
       
        dp1[2] += dp1[0];
        if(n==3)
            return max({nums[0], nums[1], nums[2]});
        
        dp1[3] += dp1[0];
        dp2[3] += dp2[1];
        if(n==4)
            return max({dp1[2], dp2[3]});
        
        dp1[4] += dp1[2];
        for(int i=5; i<n-1; i++){
            dp1[i] += max(dp1[i-2], dp1[i-3]);
        }
        
        for(int i=4; i<n; i++){
            dp2[i] += max(dp2[i-2], dp2[i-3]);
        }
        return max({dp1[n-2], dp1[n-3], dp2[n-1], dp2[n-2]});
        
    }
};
